### PR TITLE
add hic minimap2

### PR DESCRIPTION
**Pull request by yumisims**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/sanger-tol/genomeassembly/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: workflows/genomeassembly.nf [hunk 1]
```diff
@@ -91,13 +91,13 @@ workflow GENOMEASSEMBLY {
...
-    PREPARE_INPUT.out.hic.map{ meta, reads, motif -> reads }.set{ hic_reads_ch }
+    PREPARE_INPUT.out.hic.map{ meta, reads, motif, hic_aligner -> reads }.set{ hic_reads_ch }
 
     //
     // SUBWORKFLOW: GENERATE KMER DATABASE AND PROFILE MODEL
     //
-    GENOMESCOPE_MODEL( hifi_reads_ch )   
-    ch_versions = ch_versions.mix(GENOMESCOPE_MODEL.out.versions)
+    /*GENOMESCOPE_MODEL( hifi_reads_ch )   
+    ch_versions = ch_versions.mix(GENOMESCOPE_MODEL.out.versions)*/
```
Comment (ksenia-krasheninnikova) [inline workflows/genomeassembly.nf:100]: Was it commented out with the purpose of debugging and does it need uncommenting?

File: workflows/genomeassembly.nf [hunk 2]
```diff
@@ -312,24 +312,27 @@ workflow GENOMEASSEMBLY {
                        GENOMESCOPE_MODEL.out.hist,
                        GENOMESCOPE_MODEL.out.ktab
         )
-    }
+    }*/
```
Comment (ksenia-krasheninnikova) [inline workflows/genomeassembly.nf:315]: Was it commented out with the purpose of debugging and does it need uncommenting?

File: workflows/genomeassembly.nf [hunk 3]
```diff
@@ -347,7 +350,7 @@ workflow GENOMEASSEMBLY {
                        PREPARE_INPUT.out.busco,
                        GENOMESCOPE_MODEL.out.hist,
                        GENOMESCOPE_MODEL.out.ktab
-    )
+    )*/
```
Comment (ksenia-krasheninnikova) [inline workflows/genomeassembly.nf:353]: Was it commented out with the purpose of debugging and does it need uncommenting?

File: modules/local/cram_filter_minimap2_filter5end_fixmate_sort.nf [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    def args3 = task.ext.args3 ?: ''
+    def args4 = task.ext.args4 ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def VERSION = "1.15" // Staden_io versions break the pipeline
+    """
+    cram_filter -n ${from}-${to} ${cramfile} - | \\
+        samtools fastq ${args1} - |  \\
+        minimap2 -t${task.cpus} -R '${rglines}' ${args2} ${ref} - |  \\
+        ${projectDir}/bin/grep_pg.sh |  \\
+        perl ${projectDir}/bin/filter_five_end.pl |  \\
```
Comment (ksenia-krasheninnikova) [inline modules/local/cram_filter_minimap2_filter5end_fixmate_sort.nf:32]: Is there a reason why 5' filtering is not happening when [mapping is done with bwa-mem2](https://github.com/sanger-tol/genomeassembly/blob/hic_minimap2/modules/local/cram_filter_align_bwamem2_fixmate_sort.nf
) and is happening here for minimap2?

**Comment by ksenia-krasheninnikova** [comment]
@yumisims Was this version of the pipeline tested on a big dataset with `hic_aligner: minimap2` ?

File: workflows/genomeassembly.nf (hunk 1 above)
Comment (yumisims) [reply workflows/genomeassembly.nf:100]: yes, that was the purpose :) I will remove them now.

File: modules/local/cram_filter_minimap2_filter5end_fixmate_sort.nf (hunk 4 above)
Comment (yumisims) [reply modules/local/cram_filter_minimap2_filter5end_fixmate_sort.nf:32]: 5' removal happens within bwamen2 as I remember. Not sure need to look up.

**Comment by yumisims** [comment]
@ksenia-krasheninnikova not necessary to test on a bigger genome. I encountered an image downloading problem, have you seen this before?

**Comment by ksenia-krasheninnikova** [comment]
@yumisims what image exactly threw the error? Can't see it in the current CI run.

As I understand the purpose of adding minimap2 option is to be able to run the pipeline on big genomes.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -38,6 +38,9 @@ jobs:
       - name: Check out pipeline code
         uses: actions/checkout@v3
 
+      - name: Setup apptainer
```
Comment (gq1) [inline .github/workflows/ci.yml:41]: This step is already in this CI, duplicated.

Comment (gq1) [reply .github/workflows/ci.yml:41]: Check TreeVal to see how to run pipeline if you use nf-core download.

https://github.com/sanger-tol/treeval/blob/dd32db5be33201873b765306fd6f8f3eea604494/.github/workflows/ci.yml#L75

**Comment by muffato** [comment]
@yumisims : there was a problem on the `dev` in the first place. The python containers were not properly defined. https://github.com/sanger-tol/genomeassembly/commit/d72047aca168d6326d345583545aae8c9bbdc347 shows how it should be done (and no need to edit the `nf-core download` command)

File: modules/local/get_calcuts_params.nf [hunk 6]
```diff
@@ -1,10 +1,10 @@
 process GET_CALCUTS_PARAMS {
     label 'process_single'
 
-    conda "conda-forge::python"
+    conda "conda-forge::python=3.9"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/python:3' :
-        'docker.io/python:3.9' }"
```
Comment (gq1) [inline modules/local/get_calcuts_params.nf:7]: This container definition worked for nextflow run, probably not for nf-core download?

I tried to run the test on `dev` branch yesterday again without problems. 
https://github.com/sanger-tol/genomeassembly/actions/runs/8988294821

Comment (muffato) [reply modules/local/get_calcuts_params.nf:7]: I can't see any difference with the Treeval CI and container definition.
On the farm, `nf-core download` works for me.

Or maybe a version incompatibility ? (e.g. different version of nf-core or apptainer/singularity or docker)

Comment (gq1) [reply modules/local/get_calcuts_params.nf:7]: That is what we discussed yesterday in the pipeline meeting. 
Obviously `-l` flag didn't help.  Now we can try to fix the `nf-core` version as the one on local.

Comment (muffato) [reply modules/local/get_calcuts_params.nf:7]: Oki. I'll let you try that

**Comment by yumisims** [comment]
Will try when I get back.

From: Matthieu Muffato ***@***.***>
Reply to: sanger-tol/genomeassembly ***@***.***>
Date: Wednesday 29 May 2024 at 09:56
To: sanger-tol/genomeassembly ***@***.***>
Cc: Ying Sims ***@***.***>, Mention ***@***.***>
Subject: Re: [sanger-tol/genomeassembly] add hic minimap2 (PR #44) [EXT]


@muffato commented on this pull request.

________________________________

In modules/local/get_calcuts_params.nf [github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_sanger-2Dtol_genomeassembly_pull_44-23discussion-5Fr1618492295&d=DwMFaQ&c=D7ByGjS34AllFgecYw0iC6Zq7qlm8uclZFI0SqQnqBo&r=X9VPYUGy4ibZarh9Rup1XaZ62i2u-I0xOZXEy11dXDM&m=Y32RMvJHKs6Pzr27gyOwNcQBsI2Vf-sOG8ZEHvHsjsCIwxLzkOSzWinPJtGARGtG&s=6xYTGhENyr96qK19lXcCOBrc_WgrRHpnPexEF--ElJ8&e=>:

>      container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?

-        'https://depot.galaxyproject.org/singularity/python:3' :

-        'docker.io/python:3.9' }"

Oki. I'll let you try that

—
Reply to this email directly, view it on GitHub [github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_sanger-2Dtol_genomeassembly_pull_44-23discussion-5Fr1618492295&d=DwMFaQ&c=D7ByGjS34AllFgecYw0iC6Zq7qlm8uclZFI0SqQnqBo&r=X9VPYUGy4ibZarh9Rup1XaZ62i2u-I0xOZXEy11dXDM&m=Y32RMvJHKs6Pzr27gyOwNcQBsI2Vf-sOG8ZEHvHsjsCIwxLzkOSzWinPJtGARGtG&s=6xYTGhENyr96qK19lXcCOBrc_WgrRHpnPexEF--ElJ8&e=>, or unsubscribe [github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ADHNWY2GFWDDSCJGI4RUU7DZEWJ3TAVCNFSM6AAAAABH4ZA346VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAOBUG42TENBUGU&d=DwMFaQ&c=D7ByGjS34AllFgecYw0iC6Zq7qlm8uclZFI0SqQnqBo&r=X9VPYUGy4ibZarh9Rup1XaZ62i2u-I0xOZXEy11dXDM&m=Y32RMvJHKs6Pzr27gyOwNcQBsI2Vf-sOG8ZEHvHsjsCIwxLzkOSzWinPJtGARGtG&s=dVmLau18mUZGLeKhhJ7NBP2taN_hUJ_14RQBFxedYCo&e=>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

----------------------------------------------------------------------
The Wellcome Sanger Institute is operated by Genome Research Limited, a charity registered in England with number 1021457 and a company registered in England with number 2742969, whose registered office is Wellcome Sanger Institute, Wellcome Genome Campus, Hinxton, CB10 1SA.

File: assets/test_github.yaml [hunk 7]
```diff
@@ -9,6 +9,7 @@ dataset:
     reads:
       - reads: /home/runner/work/genomeassembly/genomeassembly/Undibacterium_unclassified/genomic_data/baUndUnlc1/hic-arima2/41741_2#7.sub.cram
 hic_motif: GATC,GANTC,CTNAG,TTAA
+hic_aligner: minimap2
```
Comment (ksenia-krasheninnikova) [inline assets/test_github.yaml:12]: My understanding is that the default aligner should be "bwamem2", while "minimap2" is to be used predominantly for big genomes. I'd suggest to keep this CI test configuration as well as test.yaml with the primary HiC mapping implementation i.e. "bwamem2" and add a test particularly focused on testing functionality for big genomes.

Comment (yumisims) [reply assets/test_github.yaml:12]: should I switch it back to bwamem2 for CI test for now then.

Comment (ksenia-krasheninnikova) [reply assets/test_github.yaml:12]: I would have done that! :)

**Review by ksenia-krasheninnikova — APPROVED** [review]

**Pull request merged**